ethexe/validator/src/lib.rs

```diff
@@ -16,99 +16,438 @@
+        .map(|signature| (commitments_digest, signature))
     }
 
-    pub fn push_commitments(&mut self, commitments: Vec<Commitment>) -> Result<()> {
-        for commitment in commitments {
-            match commitment {
-                Commitment::Code(code_commitment) => self.current_codes.push(code_commitment),
-                Commitment::Block(block_commitment) => self.current_blocks.push(block_commitment),
-            }
+    pub fn validate_block_commitments(
+        &mut self,
+        db: &impl BlockMetaStorage,
+        requests: impl IntoIterator<Item = BlockCommitmentValidationRequest>,
+    ) -> Result<(Digest, Signature)> {
+        let mut commitment_digests = Vec::new();
+        for request in requests.into_iter() {
+            commitment_digests.push(request.as_digest());
+            Self::validate_block_commitment(db, request)?;
         }
 
+        let commitments_digest = commitment_digests.as_digest();
+        agro::sign_commitments_digest(
+            commitments_digest,
+            &self.signer,
+            self.pub_key,
+            self.router_address,
+        )
+        .map(|signature| (commitments_digest, signature))
+    }
+
+    fn validate_code_commitment(db: &impl CodesStorage, request: CodeCommitment) -> Result<()> {
+        let CodeCommitment { id: code_id, valid } = request;
+        if db
+            .code_valid(code_id)
+            .ok_or(anyhow!("Code {code_id} is not validated by this node"))?
+            .ne(&valid)
+        {
+            return Err(anyhow!(
+                "Requested and local code validation results mismatch"
+            ));
+        }
         Ok(())
     }
 
-    pub fn publish_commitments(&mut self, sender: &mut NetworkSender) -> Result<()> {
-        let origin = self.pub_key.to_address();
+    fn validate_block_commitment(
+        db: &impl BlockMetaStorage,
+        request: BlockCommitmentValidationRequest,
+    ) -> Result<()> {
+        let BlockCommitmentValidationRequest {
+            block_hash,
+            pred_block_hash: allowed_pred_block_hash,
+            prev_commitment_hash: allowed_prev_commitment_hash,
+            transitions_digest,
+        } = request;
+
+        if db
+            .block_end_state_is_valid(block_hash)
+            .unwrap_or(false)
+            .not()
+        {
+            return Err(anyhow!(
+                "Requested block {block_hash} is not processed by this node"
+            ));
+        }
 
-        // broadcast (aggregated_code_commitments, aggregated_transitions_commitments) to the network peers
-        let commitments = (self.codes_aggregation()?, self.blocks_aggregation()?);
-        sender.publish_commitments((origin, commitments).encode());
+        if db
+            .block_outcome(block_hash)
+            .ok_or(anyhow!("Cannot get from db outcome for block {block_hash}"))?
+            .iter()
+            .map(AsDigest::as_digest)
+            .collect::<Vec<_>>()
+            .as_digest()
```

Comment:
Can be replaced with the following code, if you implement `FromIterator` for `Digest`
```suggestion
            .collect::<Digest>()
```

Reply:
Done